### ISSUE TITLE
"Server sent add for existing id" error

### ISSUE THREAD

**Issue by bradvogel** [kadirahq/fast-render#80]
Ever since introducing fastrender to my app, I'm getting "Server sent add for existing id". Any guidance on what might be the cause or how I might track it down?

**Comment by bradvogel** [comment]
@arunoda I'd like to help with a fix for this. Would you mind quickly giving me a braindump of what the timing issue might be?

**Comment by arunoda** [comment]
Hey. Thanks for that. What do you mean by the timing issue. 

This happens when you are doing a insert operation before the DDP connection comes live. 

FastRender adds data to those collection when the page loads. That's the root cause for this issue. 

As a quick fix, you can do these write ops using a method. But don't implement that method on the client. Define it only in the server.

**Comment by landland** [comment]
I get this error in a different use case, but it seems strange. If I have fast-render installed but I am not using it, meaning I do not have `fastRender: true` enabled in any of my routes, I still get this error. Does that sound possible, @arunoda ? I have to completely remove the package for the error to go away.

This is only after I switched to Cloudflare, enabled ssl and disabled websockets, and put the app behind a nginx load balancer. Even when I pause Cloudflare the issue persists. I also do not update any collections in Template rendered callbacks, but do subscribe to collections using a `Session` variable. Hope that helps.

**Comment by arunoda** [comment]
Not sure, how it relates to CF.
But your code might have some insert or update ops.

I will look for proper fix. But I can't figure out the root cause exactly.

**Comment by bradvogel** [comment]
Thanks for the workaround @arunoda !

**Comment by landland** [comment]
You're right, @arunoda. I was updating Meteor.users and that was causing the issue. Interesting how this wasn't happening without SSL/disabling websockets. The workaround of updating via server side method works. Thanks man!

**Comment by SachaG** [comment]
Just wanted to say I also run into this in Telescope from time to time.

**Comment by shtefcs** [comment]
We are facing same issue on Telescope App. 
![](http://content.screencast.com/users/shtefcs/folders/Jing/media/3536c38a-cb46-4ebc-8e1a-383c9f4bf328/2015-04-20_1406.png).

The content is loading to long, it's take few minutes to showup/render. Do u have solutions for @arunoda , cause right now all things we did on Telescope drop in water.

**Comment by arunoda** [comment]
This happens when we receive a document from server which already exists.
But it should not slow down the content loading.
If you are receiving this all the times, there seems to be something wrong.
Is this with the latest telescope?

**Comment by delgermurun** [comment]
@arunoda I'm calling method after content loaded. So it's takes so long time. If I disable fast-render it works fine.

**Comment by arunoda** [comment]
@delgermurun which method? I'd love to see a local re-production. Is this with telescope?

**Comment by SachaG** [comment]
Yes, on Telescope Meta with the latest Telescope master, I sometimes get this when submitting a comment. The comment goes through eventually, but it takes ~1 minute.

**Comment by SachaG** [comment]
Here is the code to call the comment submission method by the way: 

https://github.com/TelescopeJS/Telescope/blob/master/client/views/comments/comment_form.js

And the submitComment method: 

https://github.com/TelescopeJS/Telescope/blob/master/collections/comments.js#L206-L259

**Comment by delgermurun** [comment]
@arunoda it's on my private package and it's very simple method. Just check time and return instantly. I guess maybe, it's taking long time to connect server.

**Comment by arunoda** [comment]
Okay. May be some regression with some new changes in the core. I'll have a look at this.

**Comment by jelenajjo** [comment]
Tnx God I am not the only one getting this. My app become very slow and its take to much to load, same as above mention. If i remove fastrender app become like 5-10 slower. So really waiting for solution.

**Comment by delgermurun** [comment]
@arunoda I'm playing around this issue. I don't know how `fast-render` works. So I debug only from outside. First of all it happens very inconsistent, so it makes very hard to debug.

I reproduced on latest `devel` branch of Telescope. Logged in as an admin. I added method on server and calling it from `post_body` templates `onRendered` callback. It increases error occurrence.

I opened same post on 2 different tabs and refreshes them. So sometimes one works fine, other one raises error. Sometimes both fails or both works. How amazing is that.

I enabled log, here is some logs.

Log when throws error: https://gist.github.com/delgermurun/b96b9df994beae64b1d5
Here is log when works fine: https://gist.github.com/delgermurun/39f884fdf1834d0ba221

As you see, on those logs there is a lot difference. How that can be happen?

I'm refreshing same browser's (chrome/firefox) 2 tabs (next to each other :) ).

Seems like, if it is occurs first time, happens fairly often. What I mean is if you can produce error, it happens fairly consistent on this tab. If isn't it works fine consistently.

**Comment by voidale** [comment]
Interesting, I also get this error. I'm not sure if this fine to keep fast render and just ignore the error?

**Comment by shtefcs** [comment]
@arunoda Is there any news about this bug ?

**Comment by markoshust** [comment]
I was receiving the same error, and confirmed at least in my testing that it's only happening with https://github.com/GroundMeteor/db

**Comment by SachaG** [comment]
Could the error be related to very large documents? I'm getting it a lot when posting on http://meta.telescopeapp.org, and I'm wondering if it's because my user object contains over 700 entries for its `user.upvotedComments` property.

**Comment by arunoda** [comment]
Hmm. I'll try to replicate with that.

**Comment by arunoda** [comment]
Thanks sacha for giving me access to the telescope meta db. I found out the problem. Actually, there are a lot of stuff is happening.

But the root problem is, when there is a method call while fast-render doing it's stuff. So, the fix is to defer  all Method calls until fast-render do it's job.
I think, I can do it inside fast-render.

**Comment by epeng5** [comment]
@arunoda any updates here? I'm running into this issue as well on production.

**Comment by artiparty** [comment]
All search results of my posts look like this in google:

![pasted_image_at_2015_08_16_12_28_pm_1024](https://cloud.githubusercontent.com/assets/6714920/9292777/942512aa-4413-11e5-8516-64b5265f8cb5.png)

I believe it is related to this issue as well?

**Comment by SachaG** [comment]
Any update on this, by any chance?

**Comment by artiparty** [comment]
@SachaG once I updated Telescope version this issue solved by itself 👌

**Comment by raduchiriac** [comment]
@shtefcs what version do you use for your Telescope. Do you still see it?

@artiparty, really? it's actually the opposite. I just updated to Telescope 0.25.7 and I started seeing this error. I wonder if there will ever be a fix.

**Comment by shtefcs** [comment]
@raduchiriac I deleted that website and telescope installation. I was bit customized.

**Comment by malagant** [comment]
I have the same problem with 0.25.7.
Should I downgrade? This error prevents user from using our site. So a quick clarification would be helpful. I can also provide some more input if needed.

**Comment by SachaG** [comment]
Apparently this only happens when logged in. I'll have to dig deeper to isolate it, but maybe I'll disable FR in the meantime.

**Comment by abhaytalreja** [comment]
@SachaG - could it be cause of https://github.com/TelescopeJS/Telescope/commit/452e521ee90e2eef68e37c4829500cd2979c92fb

Based on the above comments and comparing the commits I see that's the only change in the route.js that can impact it..

**Comment by abhaytalreja** [comment]
@SachaG and @arunoda - Confirmed, If i comment this line, we are no longer getting the error. https://github.com/TelescopeJS/Telescope/blob/452e521ee90e2eef68e37c4829500cd2979c92fb/packages/telescope-posts/lib/routes.js#L18

**Comment by arunoda** [comment]
This is happens when trying to use a method when FastRendering mostly. May be you can defer it a bit.
Until the connection to Meteor is made?

We can fix this from the FastRender side as well? For now I hope you can do this.

**Comment by SachaG** [comment]
@abhaytalreja oh, good find. I don't mind delaying it, how would we do that? Would `if (Meteor.isClient) {...}` work?

**Comment by abhaytalreja** [comment]
@SachaG - Sorry for getting a little late.. i moved this increment to onRendered call of the last blaze component.

**Comment by tdarci** [comment]
@abhaytalreja... thanks for the fix. This problem has been dogging us. Where is this fixed (branch, pr, sha)? We are using 0.25.7

**Comment by abhaytalreja** [comment]
@tdarci - This issue is not resolved, i have just given the pointer and i fixed it on my local copy!

**Comment by tdarci** [comment]
@abhaytalreja ... thanks for the reply.
Is the issue due to `trackRouteEntry(params._id)` in `routes.js`?
I'd love to see your fix, if at all possible! Thanks.

**Comment by wietsevr** [comment]
@abhaytalreja Would love to see your solution as well. Thanks in advance!

**Comment by abhaytalreja** [comment]
@tdarci @wietsevr - Just delay the call `trackRouteEntry(params._id);` to this in your page load. The way i achieved it was move this line to render function of the template that loaded the last for that page.

**Comment by wietsevr** [comment]
@abhaytalreja Thanks. Works with me with `Meteor.defer()`
